core/src/cost_tracking_service.rs

```diff
@@ -0,0 +1,185 @@
+                    i64
+                ),
+                (
+                    "cost_tracker_update_elapsed",
+                    self.cost_tracker_update_elapsed as i64,
+                    i64
+                ),
+            );
+
+            *self = CostTrackingServiceStats::default();
+            self.last_print = now;
+        }
+    }
+}
+
+pub struct CommittedTransactionBatch {
+    pub transactions: Vec<Transaction>,
+    pub execution_results: Vec<TransactionExecutionResult>,
+}
+
+pub type CostTrackingReceiver = Receiver<CommittedTransactionBatch>;
+
+pub struct CostTrackingService {
+    thread_hdl: JoinHandle<()>,
+}
+
+impl CostTrackingService {
+    pub fn new(
+        exit: Arc<AtomicBool>,
+        cost_tracker: Arc<RwLock<CostTracker>>,
+        cost_tracking_receiver: CostTrackingReceiver,
+    ) -> Self {
+        let thread_hdl = Builder::new()
+            .name("solana-cost-tracking-service".to_string())
+            .spawn(move || {
+                Self::service_loop(exit, cost_tracker, cost_tracking_receiver);
+            })
+            .unwrap();
+
+        Self { thread_hdl }
+    }
+
+    pub fn join(self) -> thread::Result<()> {
+        self.thread_hdl.join()
+    }
+
+    fn service_loop(
+        exit: Arc<AtomicBool>,
+        cost_tracker: Arc<RwLock<CostTracker>>,
+        cost_tracking_receiver: CostTrackingReceiver,
+    ) {
+        let mut cost_tracking_service_stats = CostTrackingServiceStats::default();
+        let wait_timer = Duration::from_millis(10);
+
+        loop {
+            if exit.load(Ordering::Relaxed) {
+                break;
+            }
+
+            let mut cost_tracker_update_time = Measure::start("cost_tracker_update_time");
+            let mut cost_tracker_update_count = 0_u64;
+            for batch in cost_tracking_receiver.try_iter() {
+                cost_tracker_update_count += batch.transactions.len() as u64;
+                Self::process_batch(&cost_tracker, &batch);
+            }
+            cost_tracker_update_time.stop();
+            debug!("cost_update_loop cleared update channel");
+
+            cost_tracking_service_stats
+                .update(cost_tracker_update_count, cost_tracker_update_time.as_us());
+
+            thread::sleep(wait_timer);
+        }
+    }
+
+    fn process_batch(cost_tracker: &RwLock<CostTracker>, batch: &CommittedTransactionBatch) {
+        debug!(
+            "cost_tracking_service processes a batch, size {}",
+            batch.transactions.len()
+        );
+        // only track the cost of transactions that were successfully executed and committed to
+        // bank
+        for ((result, _), tx) in batch
+            .execution_results
+            .iter()
+            .zip(batch.transactions.iter())
+        {
+            if result.is_err() {
+                continue;
+            }
+            cost_tracker.write().unwrap().add_transaction_cost(tx);
```

Comment:
Not entirely sure of the future flow here, but I'm assuming the BankingStage threads will eventually be checking the cost limit via `CostTracker::try_add()`.

Isn't there the danger that if the check for the block limit is ahead of these updates (say this thread is not scheduled for a while), those banking threads will have an outdated view of the current cost?

Reply:
Also this seems prone to races with the cost tracking reset on bank boundaries: https://github.com/solana-labs/solana/blob/6188283ba61015eb1af33fe2e1d5b714f97d6926/core/src/banking_stage.rs#L430-L434

We could receive cost updates from the old bank, reset to a new bank, and then this service applies the old bank's updates on the new bank's cost model for instance. Seems like a identification like the slot of the correct cost model to update should be passed into the channel as well.

Reply:
> Not entirely sure of the future flow here, but I'm assuming the BankingStage threads will eventually be checking the cost limit via `CostTracker::try_add()`.
> 
> Isn't there the danger that if the check for the block limit is ahead of these updates (say this thread is not scheduled for a while), those banking threads will have an outdated view of the current cost?

This is a good observation. I considered it as well when making this change. My thought was this:`add_transaction_cost` to be offline, resulting banking_stage getting out dated result when calling `would_transaction_fit`

Reply:
in the commit https://github.com/solana-labs/solana/pull/18453/commits/d979e63d96b6cd142918fff08386a7a954264320

Reply:
>  however it does not fundamentally impact cost-model function, which is to estimate transaction cost in order to improve block parallelism

I was under the impression that the cost function also served the purpose of setting an upper limit on the cost of a block, and that limit could be severely undermined by bad scheduling of the cost updating service.

Would it make more sense to offload the cost calculation for the transactions to a separate stage before the transactions arrive to BankingStage? Then each transaction would have the cost information paired with it, and BankingStage just has to track the overall block cost with an atomic counter.

Reply:
This is a valid point, especially if the block limit needs to be adhered accurately. Perhaps trying to offload cost updating is a premature optimization. I am happy to discuss how that limit should be imposed, is it OK to allow a block to exceed the limit sometime so long it pays correct fee? If the block limit is to prevent DOS type attack, and preserve overall cluster health, it might be OK if a block exceeds it a bit?

Instruction's executing cost is updated in real time by replay_stage; Cost calculation is intentionally done close to when it needed to utilize the latest value.

Reply:
>  is it OK to allow a block to exceed the limit sometime so long it pays correct fee 

Hmm any scalable fee would have to be based on an accurate representation of the current cost

>If the block limit is to prevent DOS type attack, and preserve overall cluster health, it might be OK if a block exceeds it a bit?

I'm worried "A bit" is unpredictable here, as there is no predictability in thread scheduling.

Reply:
It does seem like a good idea to calculate the cost of a transaction up front before executing it though, if the idea is to save unnecessary compute